sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc

```diff
@@ -0,0 +1,390 @@
+SIMD mask parameter.
+
+[source, c++]
+----
+sycl::ext::oneapi::simd<float, 8> scale(sycl::ext::oneapi::simd<float, 8> x, float n)
+{
+  return x * n;
+}
+
+sycl::ext::oneapi::simd<float, 8> masked_scale(sycl::ext::oneapi::simd<float, 8> x, float n, sycl::ext::oneapi::simd_mask<bool, 8> mask)
+{
+  std::experimental::where(mask, x) *= n;
+  return x;
+}
+
+q.parallel_for(..., sycl::nd_item<1> it) [[sycl::reqd_sub_group_size(8)]]
+{
+  sycl::sub_group sg = it.get_sub_group();
+  float x = ...;
+  float n = ...;
+
+  // invoke SIMD function
+  // x values from each work-item are combined into a simd<float, 8>
+  float y = sycl::ext::oneapi::invoke_simd(sg, scale, x, sycl::ext::oneapi::uniform(n));
+
+  // invoke SIMD function with a mask parameter
+  // x values from each work-item be a simd<float, 8>
+  // mask values from each work-item are combined into a simd_mask<bool, 8>
+  bool mask = (it.get_local_id(0) % 2);
+  float z = sycl::ext::oneapi::invoke_simd(sg, masked_scale, x, sycl::ext::oneapi::uniform(n), mask);
+});
+----
+
+==== Execution Model
+
+Execution of a SIMD function must produce the same result as-if it had been
+executed by a single work-item executing SIMD instructions.
+
+The following are all examples of valid implementations of this model:
+
+- Mapping SIMD operations directly to SIMD instructions.
+
+- Mapping logical SIMD operations wider than the native SIMD width to multiple
+  SIMD instructions.
+
+- Mapping each element of a SIMD type to an individual work-item in a
+  sub-group, with appropriate barriers and fences generated between SIMD
+  operations.
+
+- Mapping SIMD operations to an unrolled loop executed by a single work-item.
+
+Since this execution model guarantees SIMD-like behavior, there is no need for
+the user to insert any form of explicit synchronization functions to ensure
+memory consistency across SIMD lanes.
+
+== Issues
+
+. Should we allow reference arguments?
++
+--
+*RESOLVED*: No.  Unlike pointers, it is less clear that references carry
+address space information, and users may not expect a +T&+ argument to behave
+like a pointer to the generic address space.  Users are also more likely to try
+and pass a reference to a variable in the private address space than a reference
+to a variable in the local or global address spaces, which would lead to
+undefined behavior and issues that are difficult to debug.
+--
+
+. How should returning multiple values and passing user-defined structs be handled?
++
+--
+*UNRESOLVED*: +std::tuple+ addresses both issues from a user interface perspective,
+but this aspect of the design may need to be revisited as we gain implementation experience.
+--
+
+. Should arguments require trivially copyable or device copyable?
++
+--
+*UNRESOLVED*: Device copyable permits bitwise copies of a wider range of classes, but the
+SYCL 2020 specification states that this only applies to inter-device transfers.
```

Comment:
I'd be interested to hear from @rolandschulz and @nliber on this.

The paragraphs from SYCL 2020 that this extension is referencing are these two (from https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable):
> The SYCL implementation may need to copy data between the host and a device or between two devices
> ...
> When doing an inter-device transfer of an object of type `T`, the effect of each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is the same as a bitwise copy of the object;

When the specification says "inter-device" here, I think it's classifying the host as a device.  I agree that something like "a transfer between host and device or between two devices" would by clearer.

The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary.  An implementation may still execute constructors and copy operators in the device code.

Reply:
> The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary.  An implementation may still execute constructors and copy operators in the device code.

Is it possible for user to write two different constructors: one which is invoked during host-device communication and another one which is invoked only on device side? I guess it should be possible with `__SYCL_DEVICE_ONLY__`. Do we really want to support that? Is there any particular use case where two versions of constructors are needed?

Reply:
> Is it possible for user to write two different constructors: one which is invoked during host-device communication and another one which is invoked only on device side? I guess it should be possible with __SYCL_DEVICE_ONLY__. Do we really want to support that? Is there any particular use case where two versions of constructors are needed?

I think that would be an ODR violation, and so wouldn't be allowed.

The "When doing an inter-device transfer of an object of type `T`" is what's confusing me, here, but you're right -- if there can only be one definition of each of those constructors and operators, it's unclear how a user could specialize the functions for that use-case.  But I'm going to stop guessing, and wait for the experts (@nliber and @rolandschulz) to weigh in.

Reply:
> I think that would be an ODR violation, and so wouldn't be allowed.

ODR violation apply to a single compiler. We aren't saying that this isn't allowed for split compilation (but we also not clearly saying it is allowed).

> The reason that I don't believe we can use device copyable for this extension remains unchanged, though; my reading of the paragraph above is that the constructors and operators only behave as bitwise copies when moving an object across a host-device or device-device boundary. An implementation may still execute constructors and copy operators in the device code.

I agree this is a correct reading of the wording. I don't think we spent any time wondering whether it should apply to more than device transfer because that was the use-case this was added for.

I'm not aware of any use case of device-copyable for which the copy constructor/assignment on the device does anything different from bitwise copy. And if we changed the requirement for the user to declare a type device-copyable that it doesn't do anything extra on the device we could use it here. But I think it would require a change of that paragraph to clarify it.

Reply:
Thanks, Roland.

@AlexeySachkov: Given Roland's answer, I propose that we merge this extension specification with the question about `is_device_copyable` unresolved.  I'll open a Khronos issue to discuss the intent of the specification and the need for potential clarification, and we can revisit this later.  Does that work for you?

Reply:
> @AlexeySachkov: Given Roland's answer, I propose that we merge this extension specification with the question about is_device_copyable unresolved. I'll open a Khronos issue to discuss the intent of the specification and the need for potential clarification, and we can revisit this later. Does that work for you?

Sure, no objections